mailcom/parse.py

```diff
@@ -10,86 +10,247 @@
+        self.used_last_names = {}
+
+        # forms of address that indicate last names
+        # TODO add pt and es
+        self.ln_forms_of_adress = [
+            "Madame",
+            "Monsieur",
+        ]
+
+    def init_spacy(self, language: str, model="default"):
+        if model == "default":
+            model = self.spacy_default_model_dict[language]
+        try:
+            # disable not needed components
+            self.nlp_spacy = sp.load(
+                model, exclude=["morphologizer", "attribute_ruler", "lemmatizer", "ner"]
+            )
+        except OSError:
+            raise OSError("Could not find {} in standard directory.".format(model))
+
+        self.nlp_spacy = sp.load(model)
+
+    def init_transformers(
+        self,
+        model="xlm-roberta-large-finetuned-conll03-english",
+        model_revision_number="default",
+    ):
+        # TODO: Model revision number
+
+        # ner_recognizer = pipeline("token-classification")
+        self.ner_recognizer = pipeline("token-classification", model=model)
+
+    def reset(self):
+        # reset used names for processing a new email
+        self.used_first_names.clear()
+        self.used_last_names.clear()
+
+    def get_sentences(self, input_text):
+        doc = self.nlp_spacy(input_text)
+        text_as_sents = []
+        for sent in doc.sents:
+            text_as_sents.append(str(sent))
+        return text_as_sents
+
+    def get_ner(self, sentence):
+        ner = self.ner_recognizer(sentence)
+        return ner
+
+    def pseudonymize_ne(self, ner, sentence):
+        # remove any named entities
+        entlist = []
+        new_sentence = sentence
+        # record the additional sentence length by
+        # Pseudonyms being shorter or longer as
+        # replaced names
+        additional_sentence_length = 0
+        for i in range(len(ner)):
+            entity = ner[i]
+            # check whether this entity has already been processed
+            # because it is part of a previous word
+            if entity.get("replaced", False):
+                continue
+            entity["replaced"] = True
+            ent_string = entity["entity"]  # noqa
+            # here we could check that string is "I-PER"
+            ent_conf = entity["score"]  # noqa
+            ent_position = entity["start"], entity["end"]
+            # Here we have to be careful - tokenization with
+            # transformers is quite different from spacy/stanza/flair
+            # here we get character ids
+            entlist.append(ent_position)
+            # now replace respective characters
+            # replace I-PER
+            if ent_string == "I-PER":
+                # add all following entities to this name
+                word_end = entity["end"]
+                for j in range(i, len(ner) - 1):
+                    if ner[j]["end"] != ner[j + 1]["start"]:
+                        break
+                    ner[j + 1]["replaced"] = True
```

Comment:
This approach is a lot better, although I am having troubles with the entity dict "word" entry, since it idenfities substrings of names as a seperate name. The unique list of names is then often something like this: ['▁Aga', 'the', '▁Robert'] or ['▁Le', 'fe', 'vre']. After the loop over all entities it is not reliably possible to distinguish which substring belongs to which name, or am I missing something?

Reply:
The solution to the split tokens that need to be put together into words is passing the keyword `aggregation_strategy=simple` to the transformers pipeline.